ako-gateway-api/nodes/gateway_model_rel.go

```diff
@@ -594,3 +545,147 @@ func NoOperation(namespace, name, key string) ([]string, bool) {
+				listenerHostname := akogatewayapiobjects.GatewayApiLister().GetGatewayListenerToHostname(gwListenerNsName)
+
+				hostnameMatched := false
+				for _, routeHostname := range hrObj.Spec.Hostnames {
+					// When Gateway hostname is empty, then just check validity of hostname and append it.
+					// When hostname in HTTProute has wildcard
+					// When there is exact match
+					if listenerHostname == "" || utils.CheckSubdomainOverlapping(string(routeHostname), listenerHostname) {
+						if akogatewayapilib.VerifyHostnameSubdomainMatch(string(routeHostname)) {
+							hostnameIntersection = append(hostnameIntersection, string(routeHostname))
+							hostnameMatched = true
+						}
+					}
+				}
+				// If no hostname in HTTPRoute and listener hostname is not empty, include listener hostname
+				// into list of mapped hostname between httproute and gateway (empty hostname at route and gateway)
+				if len(hrObj.Spec.Hostnames) == 0 && (listenerHostname != "" && listenerHostname != "*") {
+					hostnameIntersection = append(hostnameIntersection, string(listenerHostname))
+				}
+
+				if (hostnameMatched && !utils.HasElem(gatewayListenerList, listener)) || (len(hrObj.Spec.Hostnames) == 0 && (listenerHostname != "" && listenerHostname != "*")) {
+					gatewayListenerList = append(gatewayListenerList, listener)
+				}
+			}
+		}
+	}
+
+	if len(gatewayListenerList) > 0 {
+		gatewayToListenersMapList, _ := gatewayToListenersMap[gwNsName]
+		for _, gwListener := range gatewayListenerList {
+			if !utils.HasElem(gatewayToListenersMapList, gwListener) {
+				gatewayToListenersMapList = append(gatewayToListenersMapList, gwListener)
+			}
+		}
+		gatewayToListenersMap[gwNsName] = gatewayToListenersMapList
+	}
+	uniqueHosts := sets.NewString(hostnameIntersection...)
+	gwRouteNsName := fmt.Sprintf("%s/%s", gwNsName, routeTypeNsName)
+	akogatewayapiobjects.GatewayApiLister().UpdateGatewayRouteToHostname(gwRouteNsName, uniqueHosts.List())
+	akogatewayapiobjects.GatewayApiLister().UpdateGatewayRouteMappings(gwNsName, routeTypeNsName)
+	akogatewayapiobjects.GatewayApiLister().UpdateRouteToGatewayListenerMappings(gatewayToListenersMap[gwNsName], routeTypeNsName, gwNsName)
+	if !utils.HasElem(gwNsNameList, gwNsName) {
+		*gwNsNameList = append(*gwNsNameList, gwNsName)
+	}
+	parentNameToHostnameMap[string(parentRef.Name)] = hostnameIntersection
+}
+
+func validateReferredHTTPRoute(key, name, namespace string, allowedRoutesAll bool) ([]string, error) {
+	ns := namespace
+	if allowedRoutesAll {
+		ns = metav1.NamespaceAll
+	}
+	hrObjs, err := akogatewayapilib.AKOControlConfig().GatewayApiInformers().HTTPRouteInformer.Lister().HTTPRoutes(ns).List(labels.Set(nil).AsSelector())
+	httpRoutes := make([]*gatewayv1.HTTPRoute, 0)
+	if err != nil {
+		return nil, err
+	}
+	for _, httpRoute := range hrObjs {
+		for _, parentRef := range httpRoute.Spec.ParentRefs {
+			if parentRef.Name == gatewayv1.ObjectName(name) {
+				if IsHTTPRouteValid(key, httpRoute) {
```

Comment:
As part of this, we are checking all gateway, present as part of parent ref, valid or invalid. I think we are doing repeated validation of few things. We should work on optimising it.

Reply:
Checking if we can validate only the parent refs containing the gateway modified.  I think it will be better to add that in another PR.